File: hapsigntool/hap_sign_tool_lib/src/main/java/com/ohos/hapsigntool/zip/ZipEntryData.java
/*
 * Copyright (c) 2023-2023 Huawei Device Co., Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.ohos.hapsigntool.zip;

import com.ohos.hapsigntool.utils.FileUtils;

import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;

/**
 * resolve zip ZipEntry data
 *
 * @since 2023/12/04
 */
class ZipEntryData {
    private ZipEntryHeader zipEntryHeader;

    private long fileOffset;

    private long fileSize;

    private DataDescriptor dataDescriptor;

    private long length;

    public ZipEntryHeader getZipEntryHeader() {
        return zipEntryHeader;
    }

    /**
     * init zip entry by file
     *
     * @param file zip file
     * @param entryOffset entry start offset
     * @param compress compress file size
     * @param hasDesc has data descriptor
     * @return zip entry
     * @throws IOException read zip exception
     */
    public static ZipEntryData initZipEntry(File file, long entryOffset, long compress, boolean hasDesc)
        throws IOException {
        try (FileInputStream input = new FileInputStream(file)) {
            long offset = entryOffset;
            // read entry header by file and offset.
            byte[] headBytes = FileUtils.readInputByOffsetAndLength(input, entryOffset, ZipEntryHeader.HEADER_LENGTH);
            ZipEntryHeader entryHeader = ZipEntryHeader.initZipEntryHeader(headBytes);
            offset += ZipEntryHeader.HEADER_LENGTH;

            // read entry file name and extra by offset.
            int nameAndExtraLength = entryHeader.getFileNameLength() + entryHeader.getExtraLength();
            byte[] nameAndExtra = FileUtils.readInputByLength(input, nameAndExtraLength);
            entryHeader.setNameAndExtra(nameAndExtra);
            offset += nameAndExtraLength;

            // skip file data , save file offset and size.
            ZipEntryData entry = new ZipEntryData();
            entry.setFileOffset(offset);
            entry.setFileSize(compress);
            input.skip(compress);

            long entryLength = entryHeader.getLength() + compress;
            if (hasDesc) {
                // if entry has data descriptor, read entry data descriptor.
                byte[] desBytes = FileUtils.readInputByLength(input, DataDescriptor.DES_LENGTH);
                DataDescriptor dataDesc = DataDescriptor.initDataDescriptor(desBytes);
                entryLength += DataDescriptor.DES_LENGTH;
                entry.setDataDescriptor(dataDesc);
            }
            entry.setZipEntryHeader(entryHeader);
            entry.setLength(entryLength);
            return entry;
        }
    }

    public void setZipEntryHeader(ZipEntryHeader zipEntryHeader) {
        this.zipEntryHeader = zipEntryHeader;
    }

    public DataDescriptor getDataDescriptor() {
        return dataDescriptor;
    }

    public void setDataDescriptor(DataDescriptor dataDescriptor) {
        this.dataDescriptor = dataDescriptor;
    }

    public long getFileOffset() {
        return fileOffset;
    }

    public void setFileOffset(long fileOffset) {
        this.fileOffset = fileOffset;
    }

    public long getFileSize() {
        return fileSize;
    }

    public void setFileSize(long fileSize) {
        this.fileSize = fileSize;
    }

    public long getLength() {
        return length;
    }

    public void setLength(long length) {
        this.length = length;
    }
}

File: pc/services/zip/src/zip_entry_data.cpp
#include <fstream>
#include <iostream>
#include "zip_entry_data.h"
#include "file_utils.h"
#include "signature_tools_log.h"

using namespace OHOS::SignatureTools;

ZipEntryHeader *ZipEntryData::GetZipEntryHeader()
{
    return zipEntryHeader;
}

ZipEntryData* ZipEntryData::GetZipEntry(std::ifstream& input, long long entryOffset, long long fileSize)
{
    long long offset = entryOffset;
    // read entry header by file and offset.
    std::string retStr;
    if (FileUtils::ReadInputByOffsetAndLength(input, entryOffset, ZipEntryHeader::HEADER_LENGTH, retStr) != 0) {
        SIGNATURE_TOOLS_LOGE("ReadInputByOffsetAndLength error");
        return nullptr;
    }
    std::vector<char> headBytes(retStr.begin(), retStr.end());
    ZipEntryHeader* entryHeader = ZipEntryHeader::GetZipEntryHeader(headBytes);
    if (!entryHeader) {
        SIGNATURE_TOOLS_LOGE("entry header is nullptr.");
        return nullptr;
    }
    offset += ZipEntryHeader::HEADER_LENGTH;

    // read entry file name and extra by offset.
    if (!ReadEntryFileNameAndExtraByOffset(input, entryHeader, offset)) return nullptr;

    // skip file data , save file offset and size.
    ZipEntryData* entry = new ZipEntryData();
    entry->SetFileOffset(offset);
    entry->SetFileSize(fileSize);
    input.seekg(fileSize, std::ios::cur);
    long long entryLength = entryHeader->GetLength() + fileSize;
    short flag = entryHeader->GetFlag();
    // set desc null flag
    bool hasDesc = (flag & HAS_DATA_DESCRIPTOR_MASK) != NOT_HAS_DATA_DESCRIPTOR_FLAG;
    if (hasDesc) {
        // assuming entry has data descriptor, read entry data descriptor.
        retStr.clear();
        if (FileUtils::ReadInputByLength(input, DataDescriptor::DES_LENGTH, retStr) != 0) {
            SIGNATURE_TOOLS_LOGE("ReadInputByLength error");
            return nullptr;
        }
        std::vector<char> desBytes(retStr.begin(), retStr.end());
        DataDescriptor* dataDesc = DataDescriptor::GetDataDescriptor(desBytes);
        if (!dataDesc) {
            SIGNATURE_TOOLS_LOGE("get data descriptor failed.");
            return nullptr;
        }
        entryLength += DataDescriptor::DES_LENGTH;
        entry->SetDataDescriptor(dataDesc);
    }
    entry->SetZipEntryHeader(entryHeader);
    entry->SetLength(entryLength);
    return entry;
}

bool ZipEntryData::ReadEntryFileNameAndExtraByOffset(std::ifstream &input,
                                                     ZipEntryHeader *entryHeader,
                                                     long long &offset)
{
    if (entryHeader->GetFileNameLength() > 0) {
        std::string fileNameStr;
        if (FileUtils::ReadInputByLength(input, entryHeader->GetFileNameLength(), fileNameStr) != 0) {
            SIGNATURE_TOOLS_LOGE("Read Input By Length ERROR");
            return false;
        }
        std::vector<char> fileNameBytes(fileNameStr.begin(), fileNameStr.end());
        entryHeader->ReadFileName(fileNameBytes);
        offset += entryHeader->GetFileNameLength();
    }
    if (entryHeader->GetExtraLength() > 0) {
        std::string extraStr;
        if (FileUtils::ReadInputByLength(input, entryHeader->GetExtraLength(), extraStr) != 0) {
            SIGNATURE_TOOLS_LOGE("Read Input By Length ERROR");
            return false;
        }
        std::vector<char> extraBytes(extraStr.begin(), extraStr.end());
        entryHeader->ReadExtra(extraBytes);
        offset += entryHeader->GetExtraLength();
    }
    return true;
}

void ZipEntryData::SetZipEntryHeader(ZipEntryHeader *zipEntryHeader)
{
    this->zipEntryHeader = zipEntryHeader;
}

DataDescriptor *ZipEntryData::GetDataDescriptor()
{
    return dataDescriptor;
}

void ZipEntryData::SetDataDescriptor(DataDescriptor *dataDescriptor)
{
    this->dataDescriptor = dataDescriptor;
}

long ZipEntryData::GetFileOffset()
{
    return fileOffset;
}

void ZipEntryData::SetFileOffset(long fileOffset)
{
    this->fileOffset = fileOffset;
}

long ZipEntryData::GetFileSize()
{
    return fileSize;
}

void ZipEntryData::SetFileSize(long fileSize)
{
    this->fileSize = fileSize;
}

long ZipEntryData::GetLength()
{
    return length;
}

void ZipEntryData::SetLength(long length)
{
    this->length = length;
}